tests/js/client/shell/shell-deadlock-trx-hotbackup-cluster.js

```diff
@@ -0,0 +1,129 @@
+/* jshint globalstrict:false, strict:false, maxlen: 200 */
+/* global fail, assertTrue, assertFalse, assertEqual, assertNotUndefined, arango, print */
+
+// //////////////////////////////////////////////////////////////////////////////
+// / @brief ArangoDeadLockTests
+// /
+// /
+// / DISCLAIMER
+// /
+// / Copyright 2020 ArangoDB GmbH, Cologne, Germany
+// /
+// / Licensed under the Apache License, Version 2.0 (the "License")
+// / you may not use this file except in compliance with the License.
+// / You may obtain a copy of the License at
+// /
+// /     http://www.apache.org/licenses/LICENSE-2.0
+// /
+// / Unless required by applicable law or agreed to in writing, software
+// / distributed under the License is distributed on an "AS IS" BASIS,
+// / WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// / See the License for the specific language governing permissions and
+// / limitations under the License.
+// /
+// / @author Max Neunhoeffer
+// //////////////////////////////////////////////////////////////////////////////
+
+// tests for deadlocks in the cluster with writing trxs and hotbackup
+
+var jsunity = require('jsunity');
+var internal = require('internal');
+var arangodb = require('@arangodb');
+var db = arangodb.db;
+const isCluster = internal.isCluster();
+const time = internal.time;
+const wait = internal.wait;
+
+function trxWriteHotbackupDeadlock () {
+  'use strict';
+  var cn = 'UnitTestsDeadlockHotbackup';
+  var c = null;
+
+  return {
+
+    setUp: function () {
+      db._drop(cn);
+      c = db._create(cn, {numberOfShards: 1, replicationFactor: 2},
+                     {waitForSyncReplication: true});
+    },
+
+    tearDown: function () {
+      if (c !== null) {
+        c.drop();
+      }
+      c = null;
+      internal.wait(0);
+    },
+
+    testRunAQLWritingTransactionsDuringHotbackup: function () {
+      if (!isCluster) {
+        return true;
+      }
+
+      let start = time();
+      // This will create lots of hotbackups for approximately 60 seconds
+      let res = arango.POST_RAW("/_api/transaction",
+        { action: `function() {
+                    let console = require("console");
+                    let internal = require("internal");
+                    let wait = internal.wait;
+                    let time = internal.time;
+                    let start = time();
+                    let good = 0;
+                    let bad = 0;
+                    while (time() - start < 60) {
```

Comment:
Suggestion: I would let the two threads communicate via a special document in the collection.
For example, the async job should do the following:
```
collection.insert({ _key: "hotbackup" });
while (!collection.exists("write")) {
  internal.sleep(0.1);
}
while (collection.exists("write")) {
  do hot backups as before...
}
```
and the main thread should do:
```
post hotbackup job
collection.insert({ _key: "write" });
while (!collection.exists("hotbackup")) {
  internal.sleep(0.1);
}
start write tests as before...
collection.remove("write");
```
That way we can ensure the two threads will run overlapping, and not one after the after (which theoretically can happen).

Reply:
I find this rather theoretical. We post something asynchronously and then start immediately. If we assume that there could be a huge delay, then the same delay could happen when we check for the document and we would be back to square one. I do not think that this is worth the additional complexity. If you don't mind, I would not follow this suggestion.

Reply:
However, I have used this idea to stop the hotbackups early when all transactions are done, this decreases the test time considerably.

Reply:
Unfortunately, the whole approach does not work since we are in a transactions and thus do not see any outside change. LOL. Reverted.